sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc

```diff
@@ -0,0 +1,413 @@
+void group_store(GroupHelper gh, sycl::span<InputT, ItemsPerWorkItem> in,
+                 OutputIteratorT out_ptr,  Properties = {}); // (7)
+
+
+// Support memory function to define the needed amount of temporary memory
+// needed (name TBD)
+
+template<typename T, std::size_t ElementsPerWorkItem>
+constexpr std::size_t memory_required(sycl::memory_scope scope,
+                                      std::size_t block_size); // (8)
+
+}
+----
+
+.(6) _Preconditions_:
+* `in_ptr` must be the same for all work-items in the group.
+* `out` must be a `sycl::span` made from a pointer to the private memory space.
+
+._Constraints_:
+* Only available if `GroupHelper` is a
+work-group or sub-group or `GroupHelper` was created with a
+work-group or sub-group and some associated scratch space
+(link:https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_group_sort.asciidoc[`group_with_scratchpad`]).
+* `InputIteratorT` suppose to be a random access iterator.
+* Value type of `InputIteratorT` must be convertible to OutputT.
+
+_Effects_: Loads `ElementsPerWorkItem` elements from `in_ptr` to `out`
+using the `gh` group helper object.
+Properties may specify xref:data_placement[data placement].
+Default placement is a blocked one:
+`out[i]` = `in_ptr[gh.get_local_linear_id() * ElementsPerWorkItem + i];`
+in striped case:
+`out[i]` = `in_ptr[gh.get_local_linear_id() + gh.get_local_linear_range() * i];`
+for `i` between `0` and `ElementsPerWorkItem`.
+
+.(7) _Preconditions_:
+* `out_ptr` must be the same for all work-items in the group.
+* `in` must be a `sycl::span` made from a pointer to the private memory space.
+
+._Constraints_:
+* Only available if `GroupHelper` is a
+work-group or sub-group or `GroupHelper` was created with a
+work-group or sub-group and some associated scratch space
+(link:https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_group_sort.asciidoc[`group_with_scratchpad`]).
+* `OutputIteratorT` suppose to be a random access iterator.
+* InputT must be convertible to value type of `OutputIteratorT`.
+
+_Effects_: Stores `ElementsPerWorkItem` elements from `in` span to `out_ptr`
+using the `gh` group helper object.
+
+Properties may specify xref:data_placement[data placement].
+Default placement is a blocked one:
+`out_ptr[gh.get_local_linear_id() * ItemsPerWorkItem + i]` = `in[i];`
+in striped case:
+`out_ptr[gh.get_local_linear_id() + gh.get_local_linear_range() * i]` = `in[i];`
+for `i` between `0` and `ItemsPerWorkItem`.
+
+(8)_Effects_: Returns size of temporary memory (in bytes) that is required for
+scratch space in `GroupHelper`. Result depends on type `T`, `ElementsPerWorkItem`
+and the scope parameter: use sycl::memory_scope::work_group to get memory size
+required for each work-group; use sycl::memory_scope::sub_group to get memory
+size required for each sub-group. If other scope values are passed, behavior is
```

Comment:
> to get memory size required for each sub-group

Would user have to allocate `returned_value * number_of_sub_groups` bytes? If so, that would force usage of `[[intel::reqd_sub_group_size]]`` attribute, because otherwise that number won't be computable.

Reply:
Does group_with_scratchpad works only for sycl::group or sub_group also? If the former, then I need to double-check, maybe it is not even a use-case for sub_group to use this extra memory

Reply:
@Pennycook , do you know off the top of your head?

Reply:
I believe `group_with_scratchpad` supports any type of group.  My understanding is that it's just a way to represent a group object and a pointer simultaneously -- there shouldn't be any requirement that the pointer is in any particular memory, so a group algorithm with sub-group scope could use local or global memory.

I agree it would be easier to reason about how many sub-groups exist for kernels with `reqd_sub_group_size`.